From 3bc576a7e41c5482dbfddefeaa99e696f34de50f Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 11 Mar 2018 21:00:43 +0000
Subject: [PATCH] CONFIGURATION-688: Fixed bug related to multiple include
 files.

The locator pointing to the current configuration file was changed
when an include file was loaded. This could cause another include file
to fail.

The locator is now saved and restored after an include file has been
processed.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1826461 13f79535-47bb-0310-9956-ffa450edef68
---
 .../PropertiesConfiguration.java               | 10 +++++++++-
 .../TestPropertiesConfiguration.java           | 17 +++++++++++++++++
 .../config/testMultiInclude.properties         | 18 ++++++++++++++++++
 3 files changed, 44 insertions(+), 1 deletion(-)
 create mode 100644 src/test/resources/config/testMultiInclude.properties

diff --git a/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java b/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
index 8489b3591b..202a8be274 100644
--- a/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
@@ -1446,7 +1446,15 @@ private void loadIncludeFile(String fileName) throws ConfigurationException
 
         FileHandler fh = new FileHandler(this);
         fh.setFileLocator(locator);
-        fh.load(url);
+        FileLocator orgLocator = locator;
+        try
+        {
+            fh.load(url);
+        }
+        finally
+        {
+            locator = orgLocator; // reset locator which is changed by load
+        }
     }
 
     /**
diff --git a/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
index ded9e4582e..1332e83a3a 100644
--- a/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
@@ -233,6 +233,23 @@ public void testDisableIncludes() throws ConfigurationException,
         assertEquals("Data not loaded", PROP_VALUE, conf.getString(PROP_NAME));
     }
 
+    /**
+     * Tests whether multiple include files can be resolved.
+     */
+    @Test
+    public void testMultipleIncludeFiles() throws ConfigurationException
+    {
+        conf = new PropertiesConfiguration();
+        FileHandler handler = new FileHandler(conf);
+        handler.load(ConfigurationAssert.getTestFile("config/testMultiInclude.properties"));
+        assertEquals("Wrong top-level property", "topValue",
+                conf.getString("top"));
+        assertEquals("Wrong included property (1)", 100,
+                conf.getInt("property.c"));
+        assertEquals("Wrong included property (2)", true,
+                conf.getBoolean("include.loaded"));
+    }
+
     @Test
     public void testSetInclude() throws Exception
     {
diff --git a/src/test/resources/config/testMultiInclude.properties b/src/test/resources/config/testMultiInclude.properties
new file mode 100644
index 0000000000..cd0b66f010
--- /dev/null
+++ b/src/test/resources/config/testMultiInclude.properties
@@ -0,0 +1,18 @@
+#   Licensed to the Apache Software Foundation (ASF) under one or more
+#   contributor license agreements.  See the NOTICE file distributed with
+#   this work for additional information regarding copyright ownership.
+#   The ASF licenses this file to You under the Apache License, Version 2.0
+#   (the "License"); you may not use this file except in compliance with
+#   the License.  You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+top = topValue
+include = ../include.properties
+include = ../testEqual.properties
